ARROW-6176: [Python] Basic implementation of __arrow_ext_class__, in pure Python

balancap opened the pull request:
This is a very basic implementation of what could be `__arrow_ext_class__`, i.e. allowing extension types in PyArrow to have a custom Extension Array class (useful for adding additional logic). It is an alternative to adding dynamic attributes to `ExtensionArray` (see ARROW-8131),

Of interest @kszucs @jorisvandenbossche

jorisvandenbossche commented:
> Note, the present implementation does not keep the extension array class when serializing/deserializing to Parquet (and probably other modes of serialization). Adding this functionality would require quite a bit of additional work on the C++ codebase part 

Would this be possible from the python side by updating `pyarrow_wrap_array` to return the subclass?

balancap commented:
My bad, it's indeed simpler than I thought. I had misunderstood how the method `pyarrow_wrap_data_type` works. 

I'll add some parametrized tests for the extension type, to cover the cases where `__arrow_ext_class__` is implemented or not.

jorisvandenbossche commented on python/pyarrow/array.pxi:
```diff
@@ -2115,6 +2116,22 @@ cdef dict _array_classes = {
+        py_ext_data_type = pyarrow_wrap_data_type(sp_data_type)
+        py_ext_array_class = py_ext_data_type.__arrow_ext_class__()
+        if py_ext_array_class is NotImplementedError:
+            py_ext_array_class = ExtensionArray
```
Alternatively, that method could also return ExtensionArray by default?

balancap replied:
Make sense. I followed the existing convention, but the logic is cleaner if ExtensionArray is the default value. Now updated.

jorisvandenbossche commented on python/pyarrow/tests/test_extension_type.py:
```diff
@@ -267,14 +275,17 @@ def registered_period_type():
+def test_generic_ext_type_ipc(registered_period_type, test_ext_class):
+    period_type = PeriodType('D', test_ext_class)
     storage = pa.array([1, 2, 3, 4], pa.int64())
     arr = pa.ExtensionArray.from_storage(period_type, storage)
```
is `arr` already a PeriodArray? We should then also test that

balancap replied:
Indeed, I added a test checking this aspect

balancap commented:
@jorisvandenbossche Thanks for the feedback.

I reworked a bit the parametrization of the fixture in the tests, to behave more as expected, and such that all extension type tests are covering both cases (with/wo ext. class).

I also added a small section to the documentation `extending_types.rst`. I am afraid I am no great write though!

jorisvandenbossche commented on docs/source/python/extending_types.rst:
```diff
@@ -226,6 +226,68 @@ data type from above would look like::
+definition of the extension type.
+
+For instance, if we consider the case of a 3-dimensional vector, stored as a fixed-size
+list, where we wish to be able to extract the data as a Numpy matrix ``(N, 3)`` without
```
I would avoid using the term "matrix", as in context of numpy this can also mean something specific (i.e. the deprecated matrix class). But I think you can say "2-D numpy array"

balancap replied:
Indeed, my bad about the mixup!

balancap commented on docs/source/python/extending_types.rst:
```diff
@@ -226,6 +226,68 @@ data type from above would look like::
+type. Arrow allows to do so by adding a special method ``__arrow_ext_class__`` to the
+definition of the extension type.
+
+For instance, if we consider the case of a 3-dimensional vector, stored as a fixed-size
```
Indeed, I had not realised that it's ambiguous with the Numpy terminology. I guess I have spent too much time working with the Eigen library!

balancap commented:
Sorry @jorisvandenbossche for the delay on your comments (1 year baby having some impact on my productivity!). I have tried to use a terminology more in line with the Numpy world on the example

balancap commented:
@jorisvandenbossche Sorry to come back at you: is here anything preventing from being merged? Thanks for your help :)

balancap commented:
No worries, thanks for looking at it.
